controllers/capabilities/developeruser_controller.go

```diff
@@ -74,12 +81,61 @@ func (r *DeveloperUserReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		reqLogger.V(1).Info(string(jsonData))
 	}
 
+	// DeveloperUser has been marked for deletion
+	if developerUserCR.GetDeletionTimestamp() != nil && controllerutil.ContainsFinalizer(developerUserCR, developerUserFinalizer) {
+		err = r.removeDeveloperUserFrom3scale(developerUserCR)
+		if err != nil {
+			r.EventRecorder().Eventf(developerUserCR, corev1.EventTypeWarning, "Failed to delete developer user", "%v", err)
+			return ctrl.Result{}, err
+		}
+
+		controllerutil.RemoveFinalizer(developerUserCR, developerUserFinalizer)
+		err = r.UpdateResource(developerUserCR)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
 	// Ignore deleted resource, this can happen when foregroundDeletion is enabled
 	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
 	if developerUserCR.DeletionTimestamp != nil {
 		return ctrl.Result{}, nil
 	}
 
+	if !controllerutil.ContainsFinalizer(developerUserCR, developerUserFinalizer) {
+		controllerutil.AddFinalizer(developerUserCR, developerUserFinalizer)
+		err = r.UpdateResource(developerUserCR)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	providerAccount, err := controllerhelper.LookupProviderAccount(r.Client(), developerUserCR.GetNamespace(), developerUserCR.Spec.ProviderAccountRef, r.Logger())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Retrieve ownersReference of tenant CR that owns the DeveloperUser CR
+	tenantCR, err := controllerhelper.RetrieveTenantCR(providerAccount, r.Client(), r.Logger(), developerUserCR.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If tenant CR is found, set it's ownersReference as ownerReference in the DeveloperUser CR
+	if tenantCR != nil {
+		updated, err := controllerhelper.SetOwnersReference(developerUserCR, r.Client(), tenantCR)
```

Comment:
As said, maybe it is better to have the developer account CR as owner instead of the Tenant. wdyt?

Reply:
I was thinking of having both `Tenant` and `DeveloperAccount` as owner refs if available :thinking: 

I think there'll be an edge case of orphaned resources on cluster if we only have either or as owner ref.

The main example I'm thinking of is where `DeveloperUser` with an `providerAccountRef` to a `Tenant`, but is waiting for the linked `DeveloperAccount` to be available, and so do not have `ownerRef` to `DeveloperAccount` :thinking: . If we delete the `Tenant` now, this resouce will be left on cluster due to not being marked for deletion.

Probably not a big deal but, I think trying to have both `ownerRefs` if available should minimize orphaned resources of clusters :thinking: 

What do you think ?

Reply:
LGTM.

Both ownerrefs can be `Controller: false`. (Kubernetes does not allow multiple controller ownerreferences)

Reply:
good to know. Never explored the multi ownerref option. 

Fair enough